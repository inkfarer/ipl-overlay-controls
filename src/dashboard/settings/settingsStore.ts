import { NodeCGBrowser } from 'nodecg/browser';
import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        }
    }
});
