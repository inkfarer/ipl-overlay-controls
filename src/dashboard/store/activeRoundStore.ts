import { ActiveRound, SwapColorsInternally } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { GameWinner } from 'types/enums/gameWinner';
import { SetActiveColorRequest, SetWinnerRequest, SwapRoundColorRequest } from 'types/messages/activeRound';
import { SetRoundRequest } from 'types/messages/rounds';
import { ActiveRoundGame } from 'types/activeRoundGame';

const activeRound = nodecg.Replicant<ActiveRound>('activeRound');
const swapColorsInternally = nodecg.Replicant<SwapColorsInternally>('swapColorsInternally');

export const activeRoundReps = [ activeRound, swapColorsInternally ];

export interface ActiveRoundStore {
    activeRound: ActiveRound;
    swapColorsInternally: SwapColorsInternally;
}

export const activeRoundStore = createStore<ActiveRoundStore>({
    state: {
        activeRound: null,
        swapColorsInternally: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        setWinner(store, { winner }: { winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner });
        },
        removeWinner(): void {
            nodecg.sendMessage('removeWinner');
        },
        setActiveColor(store, request: SetActiveColorRequest): void {
            nodecg.sendMessage('setActiveColor', request);
        },
        swapColors(): void {
            swapColorsInternally.value = !swapColorsInternally.value;
        },
        setWinnerForIndex(store, { index, winner }: { index: number, winner: GameWinner }): void {
            nodecg.sendMessage('setWinner', { winner, roundIndex: index } as SetWinnerRequest);
        },
        resetActiveRound(): void {
            nodecg.sendMessage('resetActiveRound');
        },
        setActiveRound(store, request: SetRoundRequest): void {
            nodecg.sendMessage('setActiveRound', request);
        },
        swapRoundColor(store, request: SwapRoundColorRequest): void {
            nodecg.sendMessage('swapRoundColor', request);
        },
        updateActiveGames(store, games: ActiveRoundGame[]): void {
            nodecg.sendMessage('updateActiveGames', { games });
        },
        switchToNextColor(): void {
            nodecg.sendMessage('switchToNextColor');
        },
        switchToPreviousColor(): void {
            nodecg.sendMessage('switchToPreviousColor');
        },
        async getNextAndPreviousColors(): Promise<void> {
            return nodecg.sendMessage('getNextAndPreviousColors');
        }
    }
});

export const activeRoundStoreKey: InjectionKey<Store<ActiveRoundStore>> = Symbol();

export function useActiveRoundStore(): Store<ActiveRoundStore> {
    return useStore(activeRoundStoreKey);
}
