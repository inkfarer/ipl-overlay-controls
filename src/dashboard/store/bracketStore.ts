import { Bracket } from '@tourneyview/common';
import { defineStore } from 'pinia';

const bracketData = nodecg.Replicant<Bracket | undefined>('bracketData');

export const bracketReps = [ bracketData ];

export interface BracketStore {
    bracketData: Bracket | undefined;
}

export const useBracketStore = defineStore('brackets', {
    state: () => ({
        bracketData: undefined
    } as BracketStore),
    actions: {
        resetBracketData() {
            bracketData.value = undefined;
        }
    }
});
