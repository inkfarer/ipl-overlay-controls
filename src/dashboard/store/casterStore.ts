import { NodeCGBrowser } from 'nodecg/browser';
import { Caster, Casters, RadiaSettings } from 'schemas';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

const casters = nodecg.Replicant<Casters>('casters');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');

export const castersReps = [ casters, radiaSettings ];

export interface CasterStore {
    casters: Casters;
    uncommittedCasters: Casters;
    radiaSettings: RadiaSettings;
}

const defaultCaster = {
    name: 'New Caster',
    twitter: '',
    pronouns: '?/?'
};

export const useCasterStore = defineStore('casters', {
    state: () => ({
        casters: {},
        uncommittedCasters: {},
        radiaSettings: {
            enabled: null,
            guildID: null,
            updateOnImport: null
        }
    } as CasterStore),
    actions: {
        updateCaster({ id, newValue }: { id: string, newValue: Caster }): void {
            casters.value[id] = newValue;
        },
        removeUncommittedCaster(id: string): void {
            delete this.uncommittedCasters[id];
        },
        addUncommittedCaster({ id, caster }: { id: string, caster?: Caster }): void {
            this.uncommittedCasters[id] = caster ?? defaultCaster;
        },
        addDefaultCaster(): string {
            const id = generateId();
            this.addUncommittedCaster({ id });
            return id;
        },
        async saveUncommittedCaster({ id, caster }: { id: string, caster: Caster }): Promise<string> {
            const newId = await nodecg.sendMessage('saveCaster', caster);
            this.removeUncommittedCaster(id);
            return newId;
        },
        async loadCastersFromVc(): Promise<void> {
            const result = await nodecg.sendMessage('getLiveCommentators');
            if (result.extra && result.extra.length > 0) {
                for (let i = 0; i < result.extra.length; i++) {
                    const extraCaster = result.extra[i];
                    const id = extraCaster.discord_user_id;
                    delete extraCaster.discord_user_id;
                    this.addUncommittedCaster({ id, caster: extraCaster });
                }
            }
        },
        showCasters() {
            nodecg.sendMessage('mainShowCasters');
        },
        async removeCaster(id: string): Promise<void> {
            return nodecg.sendMessage('removeCaster', { id });
        }
    }
});
