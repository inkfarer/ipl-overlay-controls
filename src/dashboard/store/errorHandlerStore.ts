import { App } from 'vue';
import { generateId } from '../../helpers/generateId';
import { defineStore } from 'pinia';

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    }),
    actions: {
        removeRecentError({ key }: { key: string }): void {
            delete this.recentErrors[key];
        },
        handleError({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err);

            if (Object.keys(this.recentErrors).length >= 2) return;

            const id = generateId();
            this.recentErrors[id] = err;
            window.setTimeout(() => {
                delete this.recentErrors[id];
            }, 25000);
        }
    }
});

export function setUpErrorHandler(app: App<unknown>): void {
    const store = useErrorHandlerStore();
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info });
    };
}
