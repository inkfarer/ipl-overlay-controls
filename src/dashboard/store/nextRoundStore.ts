import { NextRound } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetNextRoundRequest } from 'types/messages/rounds';
import { BeginNextMatchRequest } from 'types/messages/activeRound';

const nextRound = nodecg.Replicant<NextRound>('nextRound');

export const nextRoundReps = [ nextRound ];

export interface NextRoundStore {
    nextRound: NextRound;
}

export const nextRoundStore = createStore<NextRoundStore>({
    state: {
        nextRound: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        },
        setShowOnStream(store, newValue: boolean) {
            nextRound.value.showOnStream = newValue;
        }
    },
    actions: {
        beginNextMatch(store, request: BeginNextMatchRequest): void {
            nodecg.sendMessage('beginNextMatch', request);
        },
        setNextRound(store, {
            teamAId,
            teamBId,
            roundId
        }: { teamAId: string, teamBId: string, roundId: string }): void {
            nodecg.sendMessage('setNextRound', { teamAId, teamBId, roundId } as SetNextRoundRequest);
        },
    }
});

export const nextRoundStoreKey: InjectionKey<Store<NextRoundStore>> = Symbol();

export function useNextRoundStore(): Store<NextRoundStore> {
    return useStore(nextRoundStoreKey);
}
