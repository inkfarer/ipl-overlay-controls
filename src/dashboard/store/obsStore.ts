import { NodeCGBrowser } from 'nodecg/browser';
import { ObsCredentials, ObsData } from 'schemas';
import { createStore, Store, useStore } from 'vuex';
import cloneDeep from 'lodash/cloneDeep';
import { InjectionKey } from 'vue';
import { SetObsDataRequest } from 'types/messages/obs';

const obsData = nodecg.Replicant<ObsData>('obsData');
const obsCredentials = nodecg.Replicant<ObsCredentials>('obsCredentials');

export const obsReps = [obsData, obsCredentials];

export interface ObsStore {
    obsData: ObsData
    obsCredentials: ObsCredentials
}

export const obsStore = createStore<ObsStore>({
    state: {
        obsData: null,
        obsCredentials: null
    },
    mutations: {
        setState(store, { name, val }: { name: string, val: unknown }): void {
            this.state[name] = cloneDeep(val);
        }
    },
    actions: {
        async connect(store, { address, password }: { address: string, password?: string }): Promise<void> {
            return nodecg.sendMessage('connectToObs', { address, password });
        },
        async setData(store, data: SetObsDataRequest): Promise<void> {
            return nodecg.sendMessage('setObsData', data);
        },
        async startGame(): Promise<void> {
            return nodecg.sendMessage('startGame');
        },
        async endGame(): Promise<void> {
            return nodecg.sendMessage('endGame');
        },
        setEnabled(store, enabled: boolean): void {
            nodecg.sendMessage('setObsSocketEnabled', enabled);
        }
    }
});

export const obsStoreKey: InjectionKey<Store<ObsStore>> = Symbol();

export function useObsStore(): Store<ObsStore> {
    return useStore(obsStoreKey);
}
