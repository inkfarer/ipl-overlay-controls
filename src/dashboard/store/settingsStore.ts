import { LastFmSettings, RadiaSettings, RuntimeConfig } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { SetGameVersionResponse } from 'types/messages/runtimeConfig';
import { defineStore } from 'pinia';
import { Locale } from 'types/enums/Locale';
import { perGameData } from '../../helpers/gameData/gameData';
import { InterfaceLocale } from 'types/enums/InterfaceLocale';

const lastFmSettings = nodecg.Replicant<LastFmSettings>('lastFmSettings');
const radiaSettings = nodecg.Replicant<RadiaSettings>('radiaSettings');
const runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');

export const settingsReps = [lastFmSettings, radiaSettings, runtimeConfig];

export interface SettingsStore {
    lastFmSettings: LastFmSettings
    radiaSettings: RadiaSettings
    runtimeConfig: RuntimeConfig
}

export const useSettingsStore = defineStore('settings', {
    state: () => ({
        lastFmSettings: {},
        radiaSettings: {
            guildID: null,
            enabled: null,
            updateOnImport: null
        },
        runtimeConfig: null
    } as SettingsStore),
    getters: {
        translatedModeName: state =>
            (mode: string) => (perGameData[state.runtimeConfig.gameVersion]
                .modes[state.runtimeConfig.locale] as Record<string, string>)[mode],
        translatedStageName: state =>
            (stage: string) => (perGameData[state.runtimeConfig.gameVersion]
                .stages[state.runtimeConfig.locale] as Record<string, string>)[stage],
    },
    actions: {
        setLastFmSettings({ newValue }: { newValue: LastFmSettings }): void {
            lastFmSettings.value = newValue;
        },
        setRadiaSettings({ newValue }: { newValue: RadiaSettings }): void {
            radiaSettings.value = newValue;
        },
        setUpdateOnImport(updateOnImport: boolean): void {
            radiaSettings.value.updateOnImport = updateOnImport;
        },
        async attemptRadiaConnection(): Promise<void> {
            return nodecg.sendMessage('retryRadiaAvailabilityCheck');
        },
        setGameVersion(newValue: GameVersion): Promise<SetGameVersionResponse> {
            return nodecg.sendMessage('setGameVersion', { version: newValue });
        },
        setLocale(newValue: Locale): Promise<void> {
            return nodecg.sendMessage('setLocale', newValue);
        },
        setInterfaceLocale(newValue: `${InterfaceLocale}`) {
            runtimeConfig.value.interfaceLocale = newValue;
        }
    }
});
