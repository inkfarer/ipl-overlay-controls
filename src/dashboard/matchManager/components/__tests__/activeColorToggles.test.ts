import ActiveColorToggles from '../activeColorToggles.vue';
import { useActiveRoundStore } from '../../../store/activeRoundStore';
import { GameWinner } from 'types/enums/gameWinner';
import { config, flushPromises, mount } from '@vue/test-utils';
import { PlayType } from 'types/enums/playType';
import { GameVersion } from 'types/enums/gameVersion';
import { createTestingPinia, TestingPinia } from '@pinia/testing';
import { useSettingsStore } from '../../../store/settingsStore';
import { IplButton } from '@iplsplatoon/vue-components';

describe('ActiveColorToggles', () => {
    let pinia: TestingPinia;

    config.global.stubs = {
        FontAwesomeIcon: true
    };

    beforeEach(() => {
        pinia = createTestingPinia();

        const activeRoundStore = useActiveRoundStore();
        activeRoundStore.getNextAndPreviousColors = jest.fn().mockResolvedValue({
            nextColor: {
                categoryName: 'Ranked Modes',
                clrA: '#FEF232',
                clrB: '#2ED2FE',
                clrNeutral: '#FD5600',
                index: 6,
                isCustom: false,
                title: 'Yellow vs Blue'
            },
            previousColor: {
                categoryName: 'Ranked Modes',
                clrA: '#04D976',
                clrB: '#D600AB',
                clrNeutral: '#D2E500',
                index: 1,
                isCustom: false,
                title: 'Green vs Magenta'
            }
        });
        activeRoundStore.$state = {
            activeRound: {
                teamA: {
                    score: 0,
                    id: '123123',
                    name: 'Team One',
                    showLogo: true,
                    players: null,
                    color: null
                },
                teamB: {
                    score: 2,
                    id: '345345',
                    name: 'Team Two (Really long name for testing ;))',
                    showLogo: false,
                    players: null,
                    color: null
                },
                activeColor: {
                    categoryName: 'Ranked Modes',
                    index: 0,
                    title: 'coolest color',
                    isCustom: false,
                    clrNeutral: '#0FAA00',
                },
                match: {
                    id: '01010',
                    name: 'Rad Match',
                    isCompleted: false,
                    type: PlayType.BEST_OF
                },
                games: [
                    {
                        winner: GameWinner.BRAVO,
                        stage: null,
                        mode: null
                    },
                    {
                        winner: GameWinner.BRAVO,
                        stage: null,
                        mode: null
                    },
                    {
                        winner: GameWinner.NO_WINNER,
                        stage: null,
                        mode: null
                    },
                ],
            },
            swapColorsInternally: false
        };

        useSettingsStore().$state = {
            // @ts-ignore
            runtimeConfig: {
                gameVersion: GameVersion.SPLATOON_2
            }
        };
    });

    it('matches snapshot', async () => {
        const wrapper = mount(ActiveColorToggles, {
            global: {
                plugins: [ pinia ]
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('matches snapshot with swapped colors', async () => {
        const store = useActiveRoundStore();
        store.swapColorsInternally = true;
        const wrapper = mount(ActiveColorToggles, {
            global: {
                plugins: [ pinia ]
            }
        });
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets color when clicking previous color toggle', async () => {
        const activeRoundStore = useActiveRoundStore();
        activeRoundStore.switchToPreviousColor = jest.fn();
        const wrapper = mount(ActiveColorToggles, {
            global: {
                plugins: [ pinia ]
            }
        });

        const toggle = wrapper.get('[data-test="color-toggle-previous"]');
        await toggle.trigger('click');

        expect(activeRoundStore.switchToPreviousColor).toHaveBeenCalled();
    });

    it('sets color when clicking next color toggle', async () => {
        const activeRoundStore = useActiveRoundStore();
        activeRoundStore.switchToNextColor = jest.fn();
        const wrapper = mount(ActiveColorToggles, {
            global: {
                plugins: [ pinia ]
            }
        });

        const toggle = wrapper.get('[data-test="color-toggle-next"]');
        await toggle.trigger('click');

        expect(activeRoundStore.switchToNextColor).toHaveBeenCalled();
    });

    it('swaps colors on swap button click', () => {
        const activeRoundStore = useActiveRoundStore();
        activeRoundStore.swapColors = jest.fn();
        const wrapper = mount(ActiveColorToggles, {
            global: {
                plugins: [ pinia ]
            }
        });

        wrapper.getComponent<typeof IplButton>('[data-test="swap-colors-button"]').vm.$emit('click');

        expect(activeRoundStore.swapColors).toHaveBeenCalled();
    });
});
