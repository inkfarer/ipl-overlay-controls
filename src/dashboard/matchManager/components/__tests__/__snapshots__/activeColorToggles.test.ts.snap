// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveColorToggles matches snapshot 1`] = `
"<div class=\\"ipl-space color-dark\\" data-v-11398c0e=\\"\\">
  <div class=\\"layout horizontal\\"><span class=\\"max-width text-small team-name wrap-anywhere\\">Team One</span><span class=\\"max-width text-small team-name wrap-anywhere text-right m-l-6\\">Team Two (Really long name for te...</span></div>
  <div class=\\"layout horizontal m-t-4\\">
    <div class=\\"color-display\\" data-test=\\"team-a-color-display\\"></div>
    <div class=\\"color-display m-l-6\\" data-test=\\"team-b-color-display\\"></div>
  </div>
  <div class=\\"layout horizontal m-t-6 color-toggle-container\\">
    <div class=\\"color-toggle layout horizontal center-vertical center-horizontal\\" data-test=\\"color-toggle-previous\\">
      <font-awesome-icon-stub icon=\\"chevron-left\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon-left\\"></font-awesome-icon-stub>
      <div class=\\"color\\" style=\\"background-color: rgb(254, 242, 50); border-color: #181e29;\\"></div>
      <div class=\\"color\\" style=\\"background-color: rgb(46, 210, 254); border-color: #181e29;\\"></div>
    </div><a class=\\"ipl-button m-l-6\\" href=\\"javascript:void(0);\\" style=\\"background-color: rgb(97, 85, 189); color: white;\\" data-test=\\"swap-colors-button\\"><span class=\\"label\\">Swap</span></a>
    <div class=\\"color-toggle layout horizontal center-vertical center-horizontal m-l-6\\" data-test=\\"color-toggle-next\\">
      <div class=\\"color\\" style=\\"background-color: rgb(4, 217, 118); border-color: #181e29;\\"></div>
      <div class=\\"color\\" style=\\"background-color: rgb(214, 0, 171); border-color: white;\\"></div>
      <font-awesome-icon-stub icon=\\"chevron-right\\" border=\\"false\\" fixedwidth=\\"false\\" listitem=\\"false\\" pulse=\\"false\\" swapopacity=\\"false\\" spin=\\"false\\" symbol=\\"false\\" inverse=\\"false\\" class=\\"icon-right\\"></font-awesome-icon-stub>
    </div>
  </div>
</div>"
`;
